src/android/post_processor_jpeg.cpp

```diff
@@ -0,0 +1,263 @@
+
+LOG_DEFINE_CATEGORY(JPEG)
+
+PostProcessorJpeg::PostProcessorJpeg(CameraDevice *const device)
+	: cameraDevice_(device)
+{
+	LOG(JPEG, Debug) << "PostProcessorJpeg::configure create!";
+}
+
+int PostProcessorJpeg::configure(const StreamConfiguration &inCfg,
+				 const StreamConfiguration &outCfg)
+{
+	LOG(JPEG, Debug) << "PostProcessorJpeg::configure start";
+
+	if (inCfg.size != outCfg.size) {
+		LOG(JPEG, Error) << "Mismatch of input and output stream sizes";
+		return -EINVAL;
+	}
+
+	if (outCfg.pixelFormat != formats::MJPEG) {
+		LOG(JPEG, Error) << "Output stream pixel format is not JPEG";
+		return -EINVAL;
+	}
+
+	streamSize_ = outCfg.size;
+
+	thumbnailer_.configure(inCfg.size, inCfg.pixelFormat);
+
+#if defined(OS_CHROMEOS)
+	encoder_ = std::make_unique<EncoderJea>();
+#else /* !defined(OS_CHROMEOS) */
+	encoder_ = std::make_unique<EncoderLibJpeg>();
+#endif
+
+	LOG(JPEG, Debug) << "PostProcessorJpeg::configure ->encoder";
+	return encoder_->configure(inCfg);
+}
+
+void PostProcessorJpeg::generateThumbnail(const FrameBuffer &source,
+					  const Size &targetSize,
+					  unsigned int quality,
+					  std::vector<unsigned char> *thumbnail)
+{
+	/* Stores the raw scaled-down thumbnail bytes. */
+	std::vector<unsigned char> rawThumbnail;
+
+	thumbnailer_.createThumbnail(source, targetSize, &rawThumbnail);
+
+	StreamConfiguration thCfg;
+	thCfg.size = targetSize;
+	thCfg.pixelFormat = thumbnailer_.pixelFormat();
+	int ret = thumbnailEncoder_.configure(thCfg);
+
+	if (!rawThumbnail.empty() && !ret) {
+		/*
+		 * \todo Avoid value-initialization of all elements of the
+		 * vector.
+		 */
+		thumbnail->resize(rawThumbnail.size());
+
+		/*
+		 * Split planes manually as the encoder expects a vector of
+		 * planes.
+		 *
+		 * \todo Pass a vector of planes directly to
+		 * Thumbnailer::createThumbnailer above and remove the manual
+		 * planes split from here.
+		 */
+		std::vector<Span<uint8_t>> thumbnailPlanes;
+		const PixelFormatInfo &formatNV12 = PixelFormatInfo::info(formats::NV12);
+		size_t yPlaneSize = formatNV12.planeSize(targetSize, 0);
+		size_t uvPlaneSize = formatNV12.planeSize(targetSize, 1);
+		thumbnailPlanes.push_back({ rawThumbnail.data(), yPlaneSize });
+		thumbnailPlanes.push_back({ rawThumbnail.data() + yPlaneSize, uvPlaneSize });
+
+		int jpeg_size = thumbnailEncoder_.encode(thumbnailPlanes,
+							 *thumbnail, {}, quality);
+		thumbnail->resize(jpeg_size);
+
+		LOG(JPEG, Debug)
+			<< "Thumbnail compress returned "
+			<< jpeg_size << " bytes";
+	}
+}
+
+void PostProcessorJpeg::process(Camera3RequestDescriptor::StreamBuffer *streamBuffer)
+{
+	ASSERT(encoder_);
+
+	const FrameBuffer &source = *streamBuffer->srcBuffer;
+	CameraBuffer *destination = streamBuffer->dstBuffer.get();
+
+	LOG(JPEG, Debug)
+		<< "PostProcessorJpeg::process streamBuffer=" << streamBuffer 
+		<< " source=" << &source
+		<<" destination=" << destination
+		<<" planes=" << destination->numPlanes();
+
+	//ASSERT(destination->numPlanes() == 1);
```

Comment:
Why did you remove the assertion ? It shouldn't trigger as the dest buffer is a single plane buffer allocated by android to host JPEG encoded data.

Reply:
No, the planes are  == 2, below is the debug output for that jpeg processor

libcamera: DEBUG JPEG post_processor_jpeg.cpp:120 PostProcessorJpeg::process streamBuffer=0xb400006f424fab10 source=0xb400006e024e5eb0 destination=0xb400006e424e2a10 **planes=2**
libcamera: DEBUG JPEG post_processor_jpeg.cpp:158 PostProcessorJpeg::process #1
ibcamera: DEBUG JPEG post_processor_jpeg.cpp:168 PostProcessorJpeg::process #2
libcamera: DEBUG JPEG post_processor_jpeg.cpp:177 PostProcessorJpeg::process #3
libcamera: DEBUG JPEG encoder_libjpeg.cpp:90 EncoderLibJpeg::configure start
libcamera: DEBUG JPEG encoder_libjpeg.cpp:110 EncoderLibJpeg::configure width=160 heigth=90 color_space=3 numPlanes=2
libcamera: DEBUG JPEG encoder_libjpeg.cpp:218 EncoderLibJpeg::encode start #2
libcamera: DEBUG JPEG encoder_libjpeg.cpp:243 JPEG Encode Starting:160x90
libcamera: DEBUG JPEG encoder_libjpeg.cpp:140 EncoderLibJpeg::compressNV start
libcamera: DEBUG JPEG post_processor_jpeg.cpp:107 Thumbnail compress returned 7600 bytes
libcamera: DEBUG JPEG post_processor_jpeg.cpp:199 PostProcessorJpeg::process #4
libcamera: DEBUG JPEG post_processor_jpeg.cpp:208 PostProcessorJpeg::process #5
libcamera: DEBUG JPEG post_processor_jpeg.cpp:218 PostProcessorJpeg::process #6
libcamera: DEBUG JPEG post_processor_jpeg.cpp:227 PostProcessorJpeg::process encode
libcamera: DEBUG JPEG encoder_libjpeg.cpp:200 EncoderLibJpeg::encode start #1
libcamera: DEBUG JPEG encoder_libjpeg.cpp:218 EncoderLibJpeg::encode start #2
libcamera: DEBUG JPEG encoder_libjpeg.cpp:243 JPEG Encode Starting:640x480
libcamera: DEBUG JPEG encoder_libjpeg.cpp:140 EncoderLibJpeg::compressNV start
libcamera: DEBUG JPEG post_processor_jpeg.cpp:235 PostProcessorJpeg::process #7
libcamera: DEBUG JPEG post_processor_jpeg.cpp:242 PostProcessorJpeg::process #8
...crash